src/Core/InstanceConfigTrait.php

```diff
@@ -68,11 +68,12 @@
      *
      * @param string|array|null $key The key to get/set, or a complete array of configs.
      * @param mixed|null $value The value to set.
-     * @param bool $merge Whether to merge or overwrite existing config, defaults to true.
+     * @param bool|string $merge 'deep' to merge recursively, 'shallow' for simple merge,
+     *   false to overwrite, defaults to 'deep'.
      * @return mixed Config value being read, or the object itself on write operations.
      * @throws \Cake\Core\Exception\Exception When trying to set a key that is invalid.
      */
-    public function config($key = null, $value = null, $merge = true)
+    public function config($key = null, $value = null, $merge = 'deep')
```

Comment:
I dunno modal options are generally a sign of a method that does too many things. We have a bunch already, bit I think it is good to think about alternatives before we add more.

Some ideas for the name are configShallow() configOverwrite() configReplace().